lib/base-compiler.js

```diff
@@ -695,6 +712,81 @@ export class BaseCompiler {
             await this.runCompiler(this.compiler.exe, newOptions, this.filename(inputFilename), execOptions));
     }
 
+    async generatePP(inputFilename, options, rawPpOptions) {
+        // -E to dump preprocessor output, remove -o so it is dumped to stdout
+        let newOptions = options.concat(['-E']);
+        if(newOptions.includes('-o')) {
+            newOptions.splice(newOptions.indexOf('-o'), 2);
+        }
+
+        const execOptions = this.getDefaultExecOptions();
+        // A higher max output is needed for when the user includes headers
+        execOptions.maxOutput = 1024 * 1024 * 1024;
+
+        let ppOptions = _.extend({
+            'filter-headers': false,
+            'clang-format': false,
+        }, rawPpOptions);
+        let result = await this.runCompilerRawOutput(this.compiler.exe, newOptions,
+                                                     this.filename(inputFilename), execOptions);
+        let numberOfLinesFiltered = 0;
+        if(ppOptions['filter-headers']) {
+            let lines = result.stdout.split('\n');
+            // Every compiler except Chibicc, as far as I've tested, outputs these line annotations
+            // Compiler test: https://godbolt.org/z/K7Pncjs4o
+            // Matching things like:
+            // # 4 "/app/example.cpp"
+            // # 11 "/usr/include/x86_64-linux-gnu/gnu/stubs.h" 2 3 4
+            // #line 1816 "C:/WinSdk/Include/10.0.18362.0/ucrt\\corecrt.h"
+            // regex test cases: https://regex101.com/r/4FgCYZ/1
+            const ppLineRe = /^\s*#\s*(?:line)?\s*\d+\s*"((?:\\"|[^"])+)"/i;
+            let region_start = -1;
+            for(let i = 0; i < lines.length; i++) {
```

Comment:
So this works, but all the complex maths and commentary relating to the update of `i` I think is subtle and potentially error prone. It's all efficient of course :) but this is javascript, so the bar's not high =)

I'd suggest a more "copy to a new output" type approach (and in a separate function to reduce the size of this, and make it testable).

To be clear, this is OK as it is but if you have any interest in changing it, I'd love it to be like:

```js
if (ppOptions['filter-headers']) {
   result.stdout = filterPP(results.stdout.split('\n')).join('\n');
}
// .... then somewhere else we can test:

filterPP(lines) {
  const result = [];
  let regionStart = -1;
  for (const line of lines) {
    // if match stuff things...
   result.append(line); // maybe?
   // something something if matches
  regionStart = result.length;
   // something something if we need to splice
  result.splice(...stuff...);
  }
  return result;
}
```